src/garage/torch/algos/maml/maml.py

```diff
@@ -0,0 +1,478 @@
+        for env in runner._sampler._vec_env.envs:
+            env.set_task(task)
+
+    @property
+    def policy(self):
+        """Current policy of the inner algorithm.
+
+        Returns:
+            garage.torch.policies.Policy: Current policy of the inner
+                algorithm.
+
+        """
+        return self._policy
+
+    @property
+    def _old_policy(self):
+        """Old policy of the inner algorithm.
+
+        Returns:
+            garage.torch.policies.Policy: Old policy of the inner algorithm.
+
+        """
+        # pylint: disable=protected-access
+        return self._inner_algo._old_policy
+
+    def _process_samples(self, itr, paths):
+        """Process sample data based on the collected paths.
+
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths
+
+        Returns:
+            MAMLTrajectoryBatch: Processed samples data.
+
+        """
+        for path in paths:
+            path['returns'] = tensor_utils.discount_cumsum(
+                path['rewards'], self._inner_algo.discount)
+
+        self._baseline.fit(paths)
+        obs, actions, rewards, valids, baselines \
+            = self._inner_algo.process_samples(itr, paths)
+        return MAMLTrajectoryBatch(obs, actions, rewards, valids, baselines)
+
+    def evaluate_performance(self, itr, all_samples, loss_before, loss_after,
+                             kl_before, kl, policy_entropy):
+        """Evaluate performance of this batch.
+
+        Args:
+            itr (int): Iteration number.
+            all_samples (list[list[MAMLTrajectoryBatch]]): Two
+                dimensional list of MAMLTrajectoryBatch of size
+                [meta_batch_size * (num_grad_updates + 1)]
+            loss_before (float): Loss before optimization step.
+            loss_after (float): Loss after optimization step.
+            kl_before (float): KL divergence before optimization step.
+            kl (float): KL divergence after optimization step.
+            policy_entropy (float): Policy entropy.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        tabular.record('Iteration', itr)
+
+        for i in range(self._num_grad_updates + 1):
+            all_rewards = [
+                path_rewards for task_samples in all_samples
+                for path_rewards in task_samples[i].rewards.numpy()
+            ]
+
+            discounted_returns = [
+                tensor_utils.discount_cumsum(path_rewards,
+                                             self._inner_algo.discount)[0]
+                for path_rewards in all_rewards
+            ]
+            undiscounted_returns = np.sum(all_rewards, axis=-1)
+            average_return = np.mean(undiscounted_returns)
+            self._episode_reward_mean[i].extend(undiscounted_returns)
+
+            with tabular.prefix('Update_{0}/'.format(i)):
```

Comment:
it is fine to have per-step logging for debugging purposes, but each meta-RL algorithm should report a single authoritative performance number for meta-train and meta-test performance.